### PR TITLE
3.0.0 Preparations

### PR DESCRIPTION
**Pull request by aurae**
These are the preparations for 3.0.0, or its first alpha at the very least, with the upcoming native Kotlin support. #355 is not part of this yet, but #350 & #349 are addressed.

We need to take care of a few TODOs for the Kotlin-based unit tests though, since we can't use the reflection-based approach to access `requestCode` and `requestPermissions` from a top-level property.

Also, the issue described in #331 doesn't seem to be fixed with the recently released Kotlin `1.1.4-3`. As stated in that issue, we should migrate over to using the new UAST-based APIs for our Lint checks, in order to make sure that Kotlin support is maximized.

Since I'll be unavailable pretty shortly, I may have to delegate the remainder of this migration to somebody else. I hope this is a nice starting point for the next milestone in this library's lifecycle!

### CODE REVIEW

**Comment by aurae** [comment]
Also, with Kotlin extension functions, our `onRequestPermissionsResult` overload is simply confusing. I strongly encourage we rethink the signature on that method! Maybe something like `validatePermissionsResult` or `processPermissionsResult`? See the Kotlin `Activity` and `Fragment` classes inside our `test` projects if you wanna know what I mean.

**Comment by aurae** [comment]
Also, this is potentially blocked by #359, since the sample projects incorrectly report our Lint rule with Kotlin 1.1.4-3. As I said over there, this is most likely happening because we're not using UAST yet.

**Comment by dbecher-ito** [comment]
Any note on when 3.0.0 will be ready? :)

**Comment by hotchemi** [comment]
@dbecher-ito after https://github.com/permissions-dispatcher/PermissionsDispatcher/pull/363 is merged~

**Comment by hotchemi** [comment]
Since @aurae is on their bicycle trip let me merge this PR and tweak a few points(now it's from his repo 's branch so I can't fix it).

File: processor/src/main/kotlin/permissions/dispatcher/processor/impl/kotlin/KotlinBaseProcessorUnit.kt [hunk 1]
```diff
@@ -159,14 +187,15 @@ abstract class BaseKtProcessorUnit : KtProcessorUnit {
...
+                builder.addStatement("%N(%N(this))", onRationale.simpleString(), permissionRequestTypeName(rpe, needsMethod))
             }
             builder.nextControlFlow("else")
         }
 
         // Add the branch for "request permission"
         ADD_WITH_CHECK_BODY_MAP[needsPermissionParameter]?.addRequestPermissionsStatement(builder, activity, requestCodeField)
-                ?: addRequestPermissionsStatement(builder, activity, permissionField, requestCodeField)
+                // TODO hardcoded "this"
+                ?: addRequestPermissionsStatement(builder, "this", permissionField, requestCodeField)
```
Comment (hotchemi) [inline processor/src/main/kotlin/permissions/dispatcher/processor/impl/kotlin/KotlinBaseProcessorUnit.kt]: Ah thx this is a bug!

**Pull request merged**